File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveItem.java
package ru.kmorozov.onedrive.client;

import com.google.api.client.util.Throwables;
import ru.kmorozov.onedrive.client.facets.FolderFacet;
import ru.kmorozov.onedrive.client.resources.Item;
import ru.kmorozov.onedrive.client.resources.ItemReference;
import ru.kmorozov.onedrive.client.serialization.JsonDateSerializer;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.text.ParseException;
import java.util.Date;

public interface OneDriveItem {
    String getId();

    boolean isDirectory();

    String getName();

    String getFullName();

    long getCrc32();

    long getSize();

    Date getCreatedDateTime();

    Date getLastModifiedDateTime();

    OneDriveItem getParent();

    FolderFacet getFolder();

    class FACTORY {

        public static OneDriveItem create(final OneDriveItem parent, final String name, final boolean isDirectory) {

            return new OneDriveItem() {
                public String getId() {
                    return null;
                }

                public boolean isDirectory() {
                    return isDirectory;
                }

                public String getName() {
                    return name;
                }

                public String getFullName() {
                    return parent.getFullName() + name + (isDirectory ? "/" : "");
                }

                @Override
                public long getCrc32() {
                    return 0L;
                }

                @Override
                public long getSize() {
                    return 0L;
                }

                @Override
                public Date getCreatedDateTime() {
                    return null;
                }

                @Override
                public Date getLastModifiedDateTime() {
                    return null;
                }

                @Override
                public OneDriveItem getParent() {
                    return parent;
                }

                @Override
                public FolderFacet getFolder() {
                    return null;
                }
            };
        }

        public static OneDriveItem create(final Item item) {
            return new OneDriveItem() {

                private final OneDriveItem parent = create(item.getParentReference());

                @Override
                public String getId() {
                    return item.getId();
                }

                @Override
                public boolean isDirectory() {
                    return null != item.getFolder();
                }

                @Override
                public String getName() {
                    return item.getName();
                }

                @Override
                public String getFullName() {
                    return parent.getFullName() + item.getName() + (isDirectory() ? "/" : "");
                }

                @Override
                public long getCrc32() {
                    return item.getFile().getHashes().getCrc32();
                }

                @Override
                public long getSize() {
                    return item.getSize();
                }

                @Override
                public Date getCreatedDateTime() {
                    try {
                        return JsonDateSerializer.INSTANCE.deserialize(item.getFileSystemInfo().getCreatedDateTime());
                    } catch (ParseException e) {
                        e.printStackTrace();

                        return new Date();
                    }
                }

                @Override
                public Date getLastModifiedDateTime() {
                    try {
                        return JsonDateSerializer.INSTANCE.deserialize(item.getFileSystemInfo().getLastModifiedDateTime());
                    } catch (ParseException e) {
                        e.printStackTrace();

                        return new Date();
                    }
                }

                @Override
                public OneDriveItem getParent() {
                    return parent;
                }

                @Override
                public FolderFacet getFolder() {
                    return item.getFolder();
                }
            };
        }

        public static OneDriveItem create(final ItemReference parent) {
            return new OneDriveItem() {
                @Override
                public String getId() {
                    return null == parent ? null : parent.getId();
                }

                @Override
                public boolean isDirectory() {
                    return true;
                }

                @Override
                public String getName() {
                    return null;
                }

                public String getFullName() {

                    if (null == parent.getPath()) {
                        return null;
                    }

                    final int index = parent.getPath().indexOf(':');

                    return URLDecoder.decode(0 < index ? parent.getPath().substring(index + 1) : parent.getPath(), StandardCharsets.UTF_8) + '/';
                }

                @Override
                public long getCrc32() {
                    return 0L;
                }

                @Override
                public long getSize() {
                    return 0L;
                }

                @Override
                public Date getCreatedDateTime() {
                    return null;
                }

                @Override
                public Date getLastModifiedDateTime() {
                    return null;
                }

                @Override
                public OneDriveItem getParent() {
                    return null;
                }

                @Override
                public FolderFacet getFolder() {
                    return null;
                }
            };
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/client/OneDriveItem.kt
package ru.kmorozov.onedrive.client

import com.google.api.client.util.Throwables
import ru.kmorozov.onedrive.client.facets.FolderFacet
import ru.kmorozov.onedrive.client.resources.Item
import ru.kmorozov.onedrive.client.resources.ItemReference
import ru.kmorozov.onedrive.client.serialization.JsonDateSerializer

import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.text.ParseException
import java.util.Date

interface OneDriveItem {
    val id: String?

    val isDirectory: Boolean

    val name: String

    val fullName: String

    val crc32: Long

    val size: Long

    val createdDateTime: Date

    val lastModifiedDateTime: Date

    val parent: OneDriveItem?

    val folder: FolderFacet?

    object FACTORY {

        fun create(parent: OneDriveItem, name: String, isDirectory: Boolean): OneDriveItem {

            return object : OneDriveItem {
                override val id: String
                    get() = ""

                override val isDirectory: Boolean
                    get() = isDirectory

                override val name: String
                    get() = name

                override val fullName: String
                    get() = parent.fullName + name + if (isDirectory) "/" else ""

                override val crc32: Long
                    get() = 0L

                override val size: Long
                    get() = 0L

                override val createdDateTime: Date
                    get() = Date()

                override val lastModifiedDateTime: Date
                    get() = Date()

                override val parent: OneDriveItem
                    get() = parent

                override val folder: FolderFacet
                    get() = FolderFacet()
            }
        }

        fun create(item: Item): OneDriveItem {
            return object : OneDriveItem {

                override val parent = create(item.parentReference)

                override val id: String
                    get() = item.id!!

                override val isDirectory: Boolean
                    get() = null != item.folder

                override val name: String
                    get() = item.name!!

                override val fullName: String
                    get() = parent.fullName + item.name + if (isDirectory) "/" else ""

                override val crc32: Long
                    get() = item.file!!.hashes!!.crc32

                override val size: Long
                    get() = item.size

                override val createdDateTime: Date
                    get() {
                        try {
                            return JsonDateSerializer.INSTANCE.deserialize(item.fileSystemInfo!!.createdDateTime!!)
                        } catch (e: ParseException) {
                            e.printStackTrace()

                            return Date()
                        }

                    }

                override val lastModifiedDateTime: Date
                    get() {
                        try {
                            return JsonDateSerializer.INSTANCE.deserialize(item.fileSystemInfo!!.lastModifiedDateTime!!)
                        } catch (e: ParseException) {
                            e.printStackTrace()

                            return Date()
                        }

                    }

                override val folder: FolderFacet
                    get() = item.folder!!
            }
        }

        fun create(parent: ItemReference?): OneDriveItem {
            return object : OneDriveItem {
                override val id: String
                    get() = parent!!.id!!

                override val isDirectory: Boolean
                    get() = true

                override val name: String
                    get() = ""

                override val fullName: String
                    get() {

                        if (null == parent!!.path) {
                            return ""
                        }

                        val index = parent.path!!.indexOf(':')

                        return URLDecoder.decode(if (0 < index) parent.path.substring(index + 1) else parent.path, StandardCharsets.UTF_8) + '/'
                    }

                override val crc32: Long
                    get() = 0L

                override val size: Long
                    get() = 0L

                override val createdDateTime: Date
                    get() = Date()

                override val lastModifiedDateTime: Date
                    get() = Date()

                override val parent: OneDriveItem?
                    get() = null

                override val folder: FolderFacet
                    get() = FolderFacet()
            }
        }
    }
}
